### ISSUE TITLE
Gutenberg does not respect `manage_categories` capability for disabling tag creation

### ISSUE THREAD

**Issue by rafaucau** [WordPress/gutenberg#50194]
### Description

I am experiencing an issue related to user roles and permissions in the Gutenberg editor. I would like to prevent editors from creating new tags while still allowing them to use the existing tags created by the administrator. 

I tried removing the `manage_categories` capability from the editor role, and while this does hide the categories and tags options from the Posts submenu, it does not fully prevent editors from creating new tags within the Gutenberg editor. Editors can still create new tags by typing in the tags input field in the post editor.

### Step-by-step reproduction instructions

1. Remove the `manage_categories` capability from a user (e.g., an editor). You can use a plugin like [User Role Editor](https://wordpress.org/plugins/user-role-editor/) to do this.
2. Log in to the account of the user with the modified permissions.

| Before  |After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/25438601/235239782-0addcf4a-4c9d-4290-808a-550ecd6f871f.png)  | ![image](https://user-images.githubusercontent.com/25438601/235239853-416e9af0-3f79-4222-bcb3-3994f2820b12.png)
3. Navigate to the post editor and create a new tag by typing it in the tags input field.
4. The new tag is created in the database, despite the user lacking the required permissions.

### Environment info

- WordPress 6.2

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by nekohayo** [comment]
Just wanted to voice my support for this, and provide some user stories / context.

I have faced the exact same problem for years, whether when dealing with my own invited authors, or when helping other organizations clean up the ontology of their blog.

Tags are waaaaay too easy to create without enforcing structure, and it inevitably rapidly becomes a mess of epic proportions. There should be a `manage_tags` permission, or indeed have the Gutenberg GUI piggyback this onto the `manage_categories` permission (at least as a stopgap measure).

No matter how many times I tell my people to not create new tags / avoid duplicates / avoid typos, they keep doing that... and I'm stuck cleaning up the mess afterwards.

**Comment by t-hamano** [comment]
To be precise, this problem is not specific to Gutenberg. Even if you remove `manage_categories`, adding tags should still be allowed in the dashboard and classic editor.

![dashboard](https://github.com/WordPress/gutenberg/assets/54422211/2e34a8d6-e53d-481b-9992-ba965337a512)
![tag](https://github.com/WordPress/gutenberg/assets/54422211/78d929e5-2d10-4496-8b5d-94ed42716325)

**Comment by webdados** [comment]
Why/how is this not fixed in 2025?

And this is not an exclusive issue of the block editor.
Even with the manage_categories capability removed from an editor, he can still easily create new category, post_tag or term from custom taxonomies on the classic editor metabox and for non-hierarchical taxonomies on the post quick edit functionality.

**Comment by timnolte** [comment]
> Why/how is this not fixed in 2025?
> 
> And this is not an exclusive issue of the block editor.
> Even with the manage_categories capability removed from an editor, he can still easily create new category, post_tag or term from custom taxonomies on the classic editor metabox and for non-hierarchical taxonomies on the post quick edit functionality.

Because no one wants to invest in the PHP parts of WordPress Core unless it serves Full Site Editing and React developers. Then again even long outstanding things that are critical to accessibility related to the Block Editor like no captions on images for the Media w/ Text block, sit dormant for years on end. If it's not new it's not a priority anymore.

**Comment by Mustafabharmal** [comment]
The issue occurred because the WordPress REST API was always (i.e.: if you have edit_terms capability) including the `wp:action-create-*` and `wp:action-assign-*` links in post responses.

The Gutenberg editor components (`FlatTermSelector` for tags) check for these action links to determine whether to allow term creation. However, since these links were always (i.e.: if you have edit_terms capability) present in the API response, the editor believed user had permission to create terms.

### Server-side Fix (`lib/rest-api.php`)
Added `gutenberg_filter_post_term_creation_links()` which:
1. Filters REST API post responses 
2. Checks the user's `manage_terms` capability (which controls term creation)
3. Removes the `wp:action-create-*` link when the user lacks this capability
4. Also removes `wp:action-assign-*` link when the user lacks `assign_terms` capability

### Client-side Fix (`packages/editor/src/components/post-taxonomies/flat-term-selector.js`)
Updated the `FlatTermSelector` component to:
1. Check `hasCreateAction` and `hasAssignAction` from the post's `_links`
2. Prevent term creation when `hasCreateAction` is false
3. Show a user-friendly error message when term creation is blocked

Will be raising the PR shortly.

**Comment by Mamaduka** [comment]
The REST API is using `edit_terms` cap for `action-create-`. IIRC, currently, if you can edit the term with the REST API, you can also create it.

Source: https://github.com/WordPress/wordpress-develop/blob/5930f083861dc48e1f4712720e62196e6cbc8793/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2324-L2332

This seems like a more general discussion about REST API I think Trac is a more suitable place for a similar discussion.

cc @TimothyBJacobs, what do you think?

**Comment by Mustafabharmal** [comment]
It looks like the issue can be resolved by changing this line:

```php
$create_cap = is_taxonomy_hierarchical( $tax->name ) ? $tax->cap->edit_terms : $tax->cap->assign_terms;
```

to:

```php
$create_cap = $tax->cap->manage_terms; // Always use manage_terms for creation
```

**Comment by TimothyBJacobs** [comment]
> It looks like the issue can be resolved by changing this line:


Yes, but that would be changing behavior for everyone and would be a BC break. 

Gutenberg and Core are behaving as designed, but this is of course a valid feature request. 

@Mamaduka I agree, this should be a core feature ticket.

**Comment by Mamaduka** [comment]
Looking at the [`create_item_permissions_check` method](https://github.com/WordPress/wordpress-develop/blob/5930f083861dc48e1f4712720e62196e6cbc8793/src/wp-includes/rest-api/endpoints/class-wp-rest-terms-controller.php#L497C18-L517), it also uses the same cap. Even if we use `canUser` selectors instead of action links, the same problem persists. Behavior can't be changed without breaking backward compatibility.

A solution here would be to also remove `edit_categories` for users who don't have `manage_categories` caps. The latter is for allowing or restricting access to admin screens and has nothing to do with term creation - https://wordpress.org/documentation/article/roles-and-capabilities/#manage_categories.

I think UI and REST API is working and following the specs.
